README
nph - PH Client and Library
===

nph is a modern client and API library for accessing PH servers.


Installation
------------

To build nph, you should be able to simply run these commands:

   ./configure
   make
   make install


Encap Package Support
---------------------

To build this software as an Encap package, you can pass the
--enable-encap option to configure.  This will be automatically
enabled if the epkg or mkencap programs are detected on the system,
but can be overridden by the --disable-encap option.

When building an Encap package, the configure script will automatically
adjust the installation prefix to use an appropriate Encap package
directory.  It does this using a heuristic algorithm which examines the
values of the ${ENCAP_SOURCE} and ${ENCAP_TARGET} environment variables
and the argument to configure's --prefix option.

If mkencap was detected on the system, it will be automatically run during
"make install".  By default, epkg will also be run, but this can be
inhibited with the --disable-epkg-install configure option.

For information on the Encap package management system, see the WSG
Encap Archive:

   http://www.encap.org/


Readline Support
----------------

The configure script will attempt to find the readline library on
your system.  The readline package is available from:

   ftp://ftp.gnu.org/gnu/readline/

If you do not have readline-4.2 or newer, you will need to specify the
--disable-readline option to configure.


Installing nph Setuid
---------------------

If you wish to use PH's email authentication mechanism, you must install
nph to be setuid-root using the --enable-setuid-nph option to configure.

SECURITY WARNING:
   Installing nph as a setuid-root binary may constitute a serious
   security hole!  While I do not know of any security holes in nph,
   I make no guaruntees about your system security.  I strongly
   recommend against using --enable-setuid-nph unless it is
   absolutely unavoidable.


Default "mailmatch" Fields
--------------------------

For PH servers that don't have a "mailmatches" entry in their siteinfo
list, libphclient will use a built-in default for the list of fields to
check when resolving email addresses.  This default list can be set by
specifying the --with-mailmatches=LIST option to configure.  (Replace
"LIST" with a ':'-delimited list of field names.)

If this option is not specified, the default list is "alias:callsign".


More Information
----------------

After you install nph, you will need to tell it what server to connect to.
This can be set in the global ${sysconfdir}/ph_server file or by setting
the environment variable ${PH_SERVER}.  Alternately, you can invoke nph
with "-s servername" to specify the server name on the commandline.

For documentation on nph or on the libphclient API, see the enclosed
manpages.  For more information on the nph package, see:

   http://www.feep.net/nph/

Both source code and Encap binary distributions of the latest version
of nph are available there.


Supported Platforms
-------------------

I develop and test nph on the following platforms:

   AIX 4.3.3
   AIX 5.1
   HP-UX 11.00
   IRIX 6.5
   RedHat Linux 7.2
   Solaris 8
   Solaris 9

The nph code should build properly on the following platforms, but I
do not actively support them:

   AIX 3.2.5
   AIX 4.2.1
   FreeBSD
   HP-UX 10.20
   Linux/libc5
   RedHat Linux 8.0
   Mac OS X
   OpenBSD
   Solaris 2.5
   Solaris 2.6
   Solaris 7

If you successfully build nph on another platform, please email me a
patch and/or configuration information.


Compatibility Code
------------------

nph depends on some library calls which are not available or not
usable on some platforms.  To accomodate these systems, I've included
a version of these calls in the lib/compat subdirectory.

I've slightly modified these functions for integration into this source
tree, but the functionality has not been modified from the original
source.  Please note that while this code should work for you, I didn't
write it, so please don't send me bug reports on it.


Author
------

Feedback and bug reports are welcome.

Mark D. Roth <roth@feep.net>



ChangeLog
nph 1.2.3 - 6/17/04
---------

- updated ph_open(3) man page
- updated ph_email_resolve(3) man page
- added PH_ERR_READONLY constant
- modified ph_change(), ph_add(), and ph_delete() to return
  PH_ERR_READONLY when server is in read-only mode
  (note that this also affects ph_passwd(), since it uses ph_change())
- modified nph "change", "add", "delete", and "edit" commands to
  check for PH_ERR_READONLY return value from libphclient
- fixed ph_email_resolve() to avoid segfault if user name is all
  punctuation or all spaces
  (thanks to Neil Rickert <rickert@cs.niu.edu> for the bug report)
- added "keyfields=" option to nph "resolve_email" command

----------------------------------------------------------------------------

nph 1.2.2 - 7/2/03
---------

- fixed ph_open() to check that result parameter of gethostbyname_r()
  and getservbyname_r() is non-NULL
  (was causing segfault under Linux for unknown hostnames; thanks to
   Damian Menscher <menscher@uiuc.edu> for the bug report)
- fixed segfault in nph "make" command

----------------------------------------------------------------------------

nph 1.2.1 - 5/9/03
---------

- minor portability fixes
- better handling of postinstall script in top-level Makefile
- fixed bug in ph_email_resolve() that caused a segfault when querying
  based on the "name" field
  (thanks to Neil Rickert <rickert@cs.niu.edu> and
   Doug Brenner <doug-brenner@uiowa.edu> for the bug reports)
- renamed MASK macro in lib/auth.c to avoid conflict with OS X header files
  (thanks to Andrew Laurence <atlauren@uci.edu> for the bug report)
- added autoconf test for _SC_GETPW_R_SIZE_MAX
  (thanks to Jon Roma <roma@uiuc.edu> and Andrew Laurence <atlauren@uci.edu>
   for the bug reports)

----------------------------------------------------------------------------

nph 1.2.0 - 11/24/02
---------

- updated README file
- added libphclient_is_thread_safe variable to libphclient API
- updated compat code
- minor portability fixes

----------------------------------------------------------------------------

nph 1.2.dev1 - 11/11/02
------------

- use poll() or select() to avoid SIGPIPE in ph_send_command()
- replaced ph_fd() function with ph_rfd() and ph_wfd()

----------------------------------------------------------------------------

nph 1.2.dev0 - 10/19/02
------------

- libphclient API changes:
  - use clearer names for constant macros in <phclient.h>
  - add prefixes to field names in all structures
  - pass application-supplied pointer to send/recv hooks
  - moved PH structure definition out of <phclient.h>
  - libphclient code is now thread-safe
- code cleanup:
  - reformatted code for readability
  - moved network line buffer code into its own module
  - moved memory management code into its own module
- packaging changes:
  - updated COPYRIGHT info
  - updated autoconf macros
  - added --with-mailmatches configure option

----------------------------------------------------------------------------

nph 1.1.8 - 6/14/02
---------

- fixed postinstall script to find nslookup in either /usr/bin or
  /usr/sbin
- updated to autoconf-2.53
- added ".PHONY" target in top-level Makefile.in to fix portability issue
  (thanks to Alex J. Lovell-Troy <alovell@uiuc.edu> for the bug report)
- removed silly extern wrapper from nph/nph.h
- avoid segfault when requesting a hidden field
- calculate query output field name width dynamically
- free fieldinfo cache in ph_login(), ph_logout(), and ph_suser() to
  accomodate hidden fields that are only visible to heros
- restructured autoconf macros

----------------------------------------------------------------------------

nph 1.1.7 - 1/30/02
---------

- fixed postinstall script to properly find domain even if hostname is
  not set to the FQDN
  (thanks to Charley LeCrone <lecrone@uiuc.edu> for the suggestion)
- use $HOME if set instead of getpwuid(getuid())
- do not use pager in non-interactive mode
- match option, siteinfo, and field names in a case-insensitive manner
- minor code cleanups
- updated to autoconf-2.52
- improved Makefile portability
- avoid segfault when "return" is the first word of a query
  (thanks to Charley LeCrone <lecrone@uiuc.edu> for the bug report)
- fixed ph_email_resolve() to properly return PH_DATAERR when the matching
  entry does not have an "email" field
  (based on patch from Doug Brenner <Doug-Brenner@uiowa.edu>)

----------------------------------------------------------------------------

nph 1.1.6 - 5/29/01
---------

- updated to use readline-4.2 API

----------------------------------------------------------------------------

nph 1.1.5 - 5/29/01
---------

- match field names in a case-insensitive manner
- updated autoconf macros and compat code
- if server doesn't support "fields" command, avoid infinite loop and
  assume default field name length of 16

----------------------------------------------------------------------------

nph 1.1.4 - 1/18/01
---------

- changed ph_serverlist_merge(), ph_email_resolve(),
  ph_advertised_email(), ph_www_resolve(), and ph_advertised_www()
  to handle PH_DATAERR when returned from ph_query()

----------------------------------------------------------------------------

nph 1.1.3 - 1/17/01
---------

- changed ph_query() to return PH_DATAERR for invalid queries
- changed ph_login() to set errno properly
- updated autoconf macros
- automatically add postinstall script when creating Encap package
- install nphrc and ph_server files from postinstall script instead of
  from make install target
- minor code cleanups

----------------------------------------------------------------------------

nph 1.1.2 - 12/13/00
---------

- fixed "set" command to properly display an error when setting a
  non-existant option
- fixed autoconf snprintf() test to make sure it NUL-terminates
- added $(DESTDIR) to Makefiles (patch from Dave Terrell <dbt@meat.net>)
- replaced $(etcdir) with $(sysconfdir) in nph/Makefile.in

lib/auth.c
/*
**  Copyright 2000-2004 University of Illinois Board of Trustees
**  Copyright 2000-2004 Mark D. Roth
**  All rights reserved.
**
**  auth.c - authentication code for libphclient
**
**  Mark D. Roth <roth@feep.net>
*/

#include <internal.h>

#include <errno.h>
#include <pwd.h>
#include <sys/types.h>

#ifdef STDC_HEADERS
# include <string.h>
# include <stdlib.h>
#endif

#ifdef HAVE_UNISTD_H
# include <unistd.h>
#endif

#ifdef HAVE_CRYPT_H
# include <crypt.h>
#endif


/* function prototypes */
typedef int (*ph_authfunc_t)(PH *, struct ph_auth *, char *, char *,
			     char *, size_t);

static int ph_passwd_auth(PH *, struct ph_auth *, char *, char *,
			  char *, size_t);
static int ph_email_auth(PH *, struct ph_auth *, char *, char *,
			 char *, size_t);


/* details of each auth type */
struct authtype
{
	int at_type;		/* auth type */
	char *at_logincmd;	/* server login command for this auth method */
	char *at_passcmd;	/* password or challenge response command */
	ph_authfunc_t at_authfunc;	/* function to modify auth token
					   (optional) */
	int at_denycode;	/* reply code sent from server to deny
				   authentication */
};
typedef struct authtype authtype_t;


static authtype_t authtypes[] = {
  { PH_AUTH_EMAIL,	"login",   "email",	ph_email_auth,	LR_NOEMAIL },
  { PH_AUTH_PASSWORD,	"login",   "answer",	ph_passwd_auth,	LR_ERROR },
  { PH_AUTH_CLEAR,	"login",   "clear",	NULL,		LR_ERROR },
  { 0,			NULL,      NULL,	NULL,		0 }
};


void
ph_free_auth(struct ph_auth *auth)
{
	if (auth->pa_alias != NULL)
		free(auth->pa_alias);
	if (auth->pa_authkey != NULL)
		free(auth->pa_authkey);
	free(auth);
}


/*
** ph_login() - login to the PH server.
** returns:
**	0				success
**	-1				error (sets errno)
**	PH_ERR_NOTLOG			authentication denied
*/
int
ph_login(PH *ph, char *alias, int authtype, void *auth)
{
	int i, code, authslot;
	char response[PH_BUF_SIZE];
	char challenge[PH_BUF_SIZE];
	struct ph_auth *authdata;
	ph_memblock_t *blockp;

#ifdef DEBUG
	printf("ph_login(): alias=\"%s\", authtype=%d, auth=\"%s\")\n",
	       alias, authtype, auth);
#endif

	/* find authtype entry */
	for (authslot = 0; authtypes[authslot].at_type != 0; authslot++)
		if (authtypes[authslot].at_type == authtype)
			break;
	if (authtypes[authslot].at_type == 0)
	{
		errno = EINVAL;
		return -1;
	}

	/* old alias no longer valid */
	if (ph->ph_auth != NULL)
		ph_free_auth(ph->ph_auth);
	ph->ph_auth = NULL;

	/*
	** free cached field info, since hidden fields are visible to
	** different users
	*/
	ph_free_fieldinfo(ph);

	if (ph_mmgr_malloc(ph->ph_mmgr, 0, 1, sizeof(struct ph_auth),
			   (ph_free_func_t)ph_free_auth, &blockp) == -1)
		return -1;
	authdata = (struct ph_auth *)ph_mmgr_ptr(blockp);
	authdata->pa_alias = strdup(alias);
	if (authdata->pa_alias == NULL)
		return -1;
	authdata->pa_authtype = authtype;

	/* send login command */
	if (ph_send_command(ph, "%s %s", authtypes[authslot].at_logincmd,
			    alias) == -1)
		return -1;

	/* read challenge */
	do
	{
		if (ph_get_response(ph, &code, challenge,
				    sizeof(challenge)) == -1)
			return -1;
	}
	while (code < LR_OK);
	if (code != LR_LOGIN)
	{
		errno = EINVAL;
		return -1;
	}

	/* determine response based on auth method */
	if (authtypes[authslot].at_authfunc != NULL)
	{
		if ((*(authtypes[authslot].at_authfunc))(ph, authdata, auth,
							 challenge, response,
							 sizeof(response)) == -1)
			return -1;
	}
	else
		strlcpy(response, auth, sizeof(response));

	/* send response */
	i = ph_send_command(ph, "%s %s", authtypes[authslot].at_passcmd,
			    response);
	if (i == -1)
		return -1;
	do
	{
		if (ph_get_response(ph, &code, NULL, 0) == -1)
			return -1;
	}
	while (code < LR_OK);

	/* successful login or login failed */
	if (code == LR_OK)
	{
		ph->ph_auth = authdata;
		ph_mmgr_free(blockp, 0);
		return 0;
	}

	if (code == authtypes[authslot].at_denycode)
		return PH_ERR_NOTLOG;

	if (code == LR_TEMP)
		errno = EAGAIN;
	else
		errno = EINVAL;
	return -1;
}


/* plugin function for email authentication */
static int
ph_email_auth(PH *ph, struct ph_auth *auth, char *password, char *challenge,
	      char *response, size_t responselen)
{
	struct passwd *pw;
#ifdef HAVE_GETPWUID_R
	struct passwd pent;
	size_t bufsize;
	ph_memblock_t *blockp;

	bufsize = (size_t)sysconf(_SC_GETPW_R_SIZE_MAX);

	if (ph_mmgr_malloc(ph->ph_mmgr, 0, 1, bufsize, free, &blockp) == -1)
		return -1;

	getpwuid_r(getuid(), &pent, ph_mmgr_ptr(blockp), bufsize, &pw);
#else /* ! HAVE_GETPWUID_R */
	pw = getpwuid(getuid());
#endif /* HAVE_GETPWUID_R */
	if (pw == NULL)
	{
#ifdef HAVE_GETPWUID_R
		ph_mmgr_free(blockp, 1);
#endif /* HAVE_GETPWUID_R */
		return -1;
	}

	strlcpy(response, pw->pw_name, responselen);

#ifdef HAVE_GETPWUID_R
	ph_mmgr_free(blockp, 1);
#endif /* HAVE_GETPWUID_R */

	return 0;
}


/* structure and definitions used for encrypted password authentication */
#define _PH_PWD_ROTORSZ         256
#define _PH_PWD_MASK            0377
struct ph_pwcrypt
{
	int n1, n2;
	char t1[_PH_PWD_ROTORSZ], t2[_PH_PWD_ROTORSZ], t3[_PH_PWD_ROTORSZ];
};


/* plugin function for encrypted password authentication */
static int
ph_passwd_auth(PH *ph, struct ph_auth *auth, char *password, char *challenge,
	       char *response, size_t responselen)
{
	int i, k, ic, seed, temp;
	unsigned random;
	char buf[13], scratch[PH_BUF_SIZE];
	char *p, *q;
	struct ph_pwcrypt *php;

#ifdef DEBUG
	printf("ph_passwd_crypt(): challenge=\"%s\", password=\"%s\"\n",
	       challenge, (password ? password : "NULL"));
#endif

	if (auth->pa_authkey == NULL)
	{
		auth->pa_authkey = (void *)malloc(sizeof(struct ph_pwcrypt));
		if (auth->pa_authkey == NULL)
			return -1;
	}

	php = (struct ph_pwcrypt *)auth->pa_authkey;

	if (password != NULL)
	{
		/* initialize arrays */
		memset(auth->pa_authkey, 0, sizeof(struct ph_pwcrypt));
		strncpy(buf, crypt(password, password), 13);
		seed = 123;
		for (i = 0; i < 13; i++)
			seed = seed * buf[i] + i;
		for (i = 0; i < _PH_PWD_ROTORSZ; i++)
			php->t1[i] = i;
		for (i = 0; i < _PH_PWD_ROTORSZ; i++)
		{
			seed = 5 * seed + buf[i % 13];
			random = seed % 65521;
			k = _PH_PWD_ROTORSZ - 1 - i;
			ic = (random & _PH_PWD_MASK) % (k + 1);
			random >>= 8;
			temp = php->t1[k];
			php->t1[k] = php->t1[ic];
			php->t1[ic] = temp;
			if (php->t3[k] != 0)
				continue;
			ic = (random & _PH_PWD_MASK) % k;
			while (php->t3[ic] != 0)
				ic = (ic + 1) % k;
			php->t3[k] = ic;
			php->t3[ic] = k;
		}
		for (i = 0; i < _PH_PWD_ROTORSZ; i++)
			php->t2[php->t1[i] & _PH_PWD_MASK] = i;
	}

	/* decode challenge into scratch */
	for (p = challenge, q = scratch;
	     (q - scratch < sizeof(scratch)) && *p;
	     p++)
	{
		*q++ = php->t2[(php->t3[(php->t1[(*p + php->n1) & _PH_PWD_MASK] + php->n2) & _PH_PWD_MASK] - php->n2) & _PH_PWD_MASK] - php->n1;
		php->n1++;
		if (php->n1 == _PH_PWD_ROTORSZ)
		{
			php->n1 = 0;
			php->n2++;
			if (php->n2 == _PH_PWD_ROTORSZ)
				php->n2 = 0;
		}
	}

	/* encode as ASCII into response buffer */
	k = q - scratch;
	p = response;
	q = scratch;
	*p++ = (k & 077) + '#';
	for (; (p - response < responselen) && (q - scratch < k); q += 3)
	{
		*p++ = ((*q >> 2) & 077) + '#';
		*p++ = ((((*q << 4) & 060) | ((q[1] >> 4) & 017)) & 077) + '#';
		*p++ = ((((q[1] << 2) & 074) | ((q[2] >> 6) & 03)) & 077) + '#';
		*p++ = ((q[2] & 077) & 077) + '#';
	}
	*p = '\0';

	return 0;
}


/*
** ph_logout() - logout of the PH server.
** returns:
**      0                               success
**      -1                              error (sets errno)
**      PH_ERR_NOTLOG                       not logged in
*/
int
ph_logout(PH *ph)
{
	int code;

	if (ph->ph_auth == NULL)
		return PH_ERR_NOTLOG;

	/* old alias now invalid */
	ph_free_auth(ph->ph_auth);
	ph->ph_auth = NULL;

	/*
	** free cached field info, since hidden fields are visible to
	** different users
	*/
	ph_free_fieldinfo(ph);

	if (ph_send_command(ph, "logout") == -1)
		return -1;

	do
	{
		if (ph_get_response(ph, &code, NULL, 0) == -1)
			return -1;
	}
	while (code < LR_OK);

	if (code == LR_ERROR)
		return PH_ERR_NOTLOG;
	if (code == LR_OK)
		return 0;

	errno = EINVAL;
	return -1;
}


/*
** ph_suser() - assume another user's identity (heros only).
** returns:
**      0                               success
**      -1                              error (sets errno)
**      PH_ERR_NOTHERO                      not a hero
**      PH_ERR_NOTLOG                       not logged in
*/
int
ph_suser(PH *ph, char *alias)
{
	int code;

	if (ph->ph_auth == NULL)
		return PH_ERR_NOTLOG;

	if (ph_send_command(ph, "suser %s", alias) == -1)
		return -1;

	do
	{
		if (ph_get_response(ph, &code, NULL, 0) == -1)
			return -1;
	}
	while (code < LR_OK);

	if (code == LR_AUTH)
		return PH_ERR_NOTHERO;
	if (code == LR_OK)
	{
		/*
		** free cached field info, since hidden fields are visible to
		** different users
		*/
		ph_free_fieldinfo(ph);

		if (ph->ph_auth->pa_alias != NULL)
			free(ph->ph_auth->pa_alias);
		ph->ph_auth->pa_alias = strdup(alias);
		if (ph->ph_auth->pa_alias != NULL)
			return -1;
		return 0;
	}

	errno = EINVAL;
	return -1;
}


/*
** ph_passwd() - change the current user's PH password.
** returns:
**	0				success
**	-1				error (sets errno)
**	PH_ERR_NOTLOG			not logged in
**	(any value returned by ph_change())
*/
int
ph_passwd(PH *ph, char *newpass)
{
	struct ph_fieldselector selector[2];
	struct ph_fieldselector changes[2];
	char encrypted_passwd[PH_BUF_SIZE];
	int i;

#ifdef DEBUG
	printf("==> ph_passwd(newpass=\"%s\")\n", newpass);
#endif

	/* must be logged in to change password */
	if (ph->ph_auth == NULL)
		return PH_ERR_NOTLOG;

	selector[0].pfs_field = "alias";
	selector[0].pfs_operation = '=';
	selector[0].pfs_value = ph->ph_auth->pa_alias;
	memset(&(selector[1]), 0, sizeof(struct ph_fieldselector));

	/* generate the encrypted password */
	ph_passwd_auth(ph, ph->ph_auth, NULL, newpass,
		       encrypted_passwd, sizeof(encrypted_passwd));
#ifdef DEBUG
	printf("encrypted_passwd = \"%s\"\n", encrypted_passwd);
#endif
	changes[0].pfs_field = "password";
	changes[0].pfs_operation = '=';
	changes[0].pfs_value = encrypted_passwd;
	memset(&(changes[1]), 0, sizeof(struct ph_fieldselector));

#ifdef DEBUG
	puts("calling ph_change()...");
#endif

	i = ph_change(ph, selector, changes, 0);
	if (i == 1)
		return 0;

	return i;
}


/*
** ph_whoami() - determine the currently logged-in alias
** returns:
**      pointer to the current alias    success
**      NULL                            not logged in
*/
char *
ph_whoami(PH *ph)
{
	return (ph->ph_auth != NULL
		? ph->ph_auth->pa_alias
		: NULL);
}


